File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def fn(node):
            if not node:
                return
            
            if node.val == p.val:
                return p
            
            if node.val == q.val:
                return q
            
            left = fn(node.left)
            right = fn(node.right)

            if left and right:
                return node
            
            if left:
                return left
            
            if right:
                return right

            return None
        return fn(root)

File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode fn(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        }
        TreeNode left = fn(root.left, p, q);
        TreeNode right = fn(root.right, p, q);
        if (root.val == p.val) {
            return p;
        }
        if (root.val == q.val) {
            return q;
        }
        if (left != null && right != null) {
            return root;
        }
        // if (left == null && right == null) {
        //     return null;
        // }
        if (left != null) {
            return left;
        }
        if (right != null) {
            return right;
        }
        return null;
    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return fn(root,p,q);
    }
}